// gifted-charts-core/dist/BarChart/RenderStackBars.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import { useState } from 'react';
import { BarDefaults } from '../utils/constants';
export var useRenderStackBars = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    var item = props.item, index = props.index, _k = props.containerHeight, containerHeight = _k === void 0 ? 200 : _k, containerHeightIncludingBelowXAxis = props.containerHeightIncludingBelowXAxis, maxValue = props.maxValue, propSpacing = props.propSpacing, initialSpacing = props.initialSpacing, stackData = props.stackData, isAnimated = props.isAnimated, xAxisThickness = props.xAxisThickness, renderTooltip = props.renderTooltip, leftShiftForTooltip = props.leftShiftForTooltip, leftShiftForLastIndexTooltip = props.leftShiftForLastIndexTooltip, autoCenterTooltip = props.autoCenterTooltip, horizontal = props.horizontal, stepHeight = props.stepHeight, stepValue = props.stepValue, negativeStepHeight = props.negativeStepHeight, negativeStepValue = props.negativeStepValue, secondaryStepHeight = props.secondaryStepHeight, secondaryStepValue = props.secondaryStepValue, secondaryNegativeStepHeight = props.secondaryNegativeStepHeight, secondaryNegativeStepValue = props.secondaryNegativeStepValue;
    var containsNegativeValue = item.stacks.some(function (item) { return item.value < 0; });
    var anyStackContainsNegativeValue = stackData.some(function (item) {
        return item.stacks.some(function (stack) { return stack.value < 0; });
    });
    var noAnimation = containsNegativeValue || !isAnimated;
    var localBarInnerComponent = (_a = item.barInnerComponent) !== null && _a !== void 0 ? _a : props.barInnerComponent;
    var borderRadius = item.borderRadius, borderTopLeftRadius = item.borderTopLeftRadius, borderTopRightRadius = item.borderTopRightRadius, borderBottomLeftRadius = item.borderBottomLeftRadius, borderBottomRightRadius = item.borderBottomRightRadius;
    var leftSpacing = initialSpacing;
    for (var i = 0; i < index; i++) {
        leftSpacing +=
            ((_c = (_b = stackData[i].spacing) !== null && _b !== void 0 ? _b : propSpacing) !== null && _c !== void 0 ? _c : 0) +
                ((_e = (_d = stackData[i].stacks[0].barWidth) !== null && _d !== void 0 ? _d : props.barWidth) !== null && _e !== void 0 ? _e : 30);
    }
    var disablePress = (_f = props.disablePress) !== null && _f !== void 0 ? _f : false;
    var heightFactor = item.isSecondary
        ? secondaryStepHeight / secondaryStepValue
        : stepHeight / stepValue;
    var negativeHeightFactor = item.isSecondary
        ? secondaryNegativeStepHeight / secondaryNegativeStepValue
        : negativeStepHeight / negativeStepValue;
    var totalHeight = props.item.stacks.reduce(function (acc, stack) {
        return acc +
            Math.abs(stack.value) *
                (stack.value < 0 ? negativeHeightFactor : heightFactor);
    }, 0);
    var _l = __read(useState(noAnimation ? totalHeight : 1), 2), height = _l[0], setHeight = _l[1];
    var getBarHeight = function (value, marginBottom) {
        return (Math.abs(value) * (value < 0 ? negativeHeightFactor : heightFactor) -
            (marginBottom !== null && marginBottom !== void 0 ? marginBottom : 0));
    };
    var getPosition = function (index, height) {
        /* Returns bottom position for stack item
               negative values are below origin (-> negative position) */
        var itemValue = item.stacks[index].value;
        var isNegative = itemValue <= 0;
        var position = isNegative ? -(height || 0) - xAxisThickness : 0;
        for (var i = 0; i < index; i++) {
            var valueOnIndex = item.stacks[i].value;
            if (isNegative && valueOnIndex <= 0) {
                position +=
                    (valueOnIndex * (containerHeight !== null && containerHeight !== void 0 ? containerHeight : 200)) / (maxValue || 200);
            }
            else if (!isNegative && valueOnIndex >= 0) {
                position +=
                    (valueOnIndex * (containerHeight !== null && containerHeight !== void 0 ? containerHeight : 200)) / (maxValue || 200);
            }
        }
        return position;
    };
    var getLowestPosition = function () {
        var _a;
        return (((_a = item.stacks
            .map(function (_, index) {
            return getPosition(index, getBarHeight(_.value, _.marginBottom));
        })
            .sort(function (a, b) { return a - b; })) === null || _a === void 0 ? void 0 : _a[0]) || 0);
    };
    var getPositiveBoxesHeightSum = function () {
        var sum = 0;
        item.stacks.forEach(function (_, ind) {
            var value = _.value;
            sum += value > 0 ? getBarHeight(value, _.marginBottom) : 0;
        });
        return sum;
    };
    var lowestBarPosition = getLowestPosition();
    var getStackBorderRadii = function (item, index) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
        var stackItem = item.stacks[index];
        var borderRadii = {
            borderTopLeftRadius: (_d = (_c = (_b = (_a = stackItem.borderTopLeftRadius) !== null && _a !== void 0 ? _a : stackItem.borderRadius) !== null && _b !== void 0 ? _b : props.barBorderTopLeftRadius) !== null && _c !== void 0 ? _c : props.barBorderRadius) !== null && _d !== void 0 ? _d : 0,
            borderTopRightRadius: (_h = (_g = (_f = (_e = stackItem.borderTopRightRadius) !== null && _e !== void 0 ? _e : stackItem.borderRadius) !== null && _f !== void 0 ? _f : props.barBorderTopRightRadius) !== null && _g !== void 0 ? _g : props.barBorderRadius) !== null && _h !== void 0 ? _h : 0,
            borderBottomLeftRadius: (_m = (_l = (_k = (_j = stackItem.borderBottomLeftRadius) !== null && _j !== void 0 ? _j : stackItem.borderRadius) !== null && _k !== void 0 ? _k : props.barBorderBottomLeftRadius) !== null && _l !== void 0 ? _l : props.barBorderRadius) !== null && _m !== void 0 ? _m : 0,
            borderBottomRightRadius: (_r = (_q = (_p = (_o = stackItem.borderBottomRightRadius) !== null && _o !== void 0 ? _o : stackItem.borderRadius) !== null && _p !== void 0 ? _p : props.barBorderBottomRightRadius) !== null && _q !== void 0 ? _q : props.barBorderRadius) !== null && _r !== void 0 ? _r : 0
        };
        return borderRadii;
    };
    var chartHeightBelowXAxis = containerHeightIncludingBelowXAxis - containerHeight;
    var barHeight = totalHeight;
    var getBotomForTooltip = function () {
        var _a;
        if (anyStackContainsNegativeValue) {
            // If any of the stacks has a negative value
            if (item.stacks.some(function (stack) { return stack.value > 0; })) {
                // if a box in the current stack has a +ve value
                return chartHeightBelowXAxis + ((_a = getPositiveBoxesHeightSum()) !== null && _a !== void 0 ? _a : 0) - 8;
            }
            else
                return chartHeightBelowXAxis - barHeight - 34;
        }
        else {
            return barHeight + 2;
        }
    };
    var tooltipProps = {
        barHeight: barHeight,
        barWidth: item.barWidth || props.barWidth || 30,
        item: item,
        index: index,
        isLast: index === stackData.length - 1,
        leftSpacing: leftSpacing,
        leftShiftForLastIndexTooltip: leftShiftForLastIndexTooltip,
        leftShiftForTooltip: (_h = (_g = item.leftShiftForTooltip) !== null && _g !== void 0 ? _g : leftShiftForTooltip) !== null && _h !== void 0 ? _h : 0,
        renderTooltip: renderTooltip,
        autoCenterTooltip: autoCenterTooltip,
        horizontal: horizontal,
        bottom: getBotomForTooltip()
    };
    return {
        containsNegativeValue: containsNegativeValue,
        noAnimation: noAnimation,
        localBarInnerComponent: localBarInnerComponent,
        borderRadius: borderRadius,
        borderTopLeftRadius: borderTopLeftRadius,
        borderTopRightRadius: borderTopRightRadius,
        borderBottomLeftRadius: borderBottomLeftRadius,
        borderBottomRightRadius: borderBottomRightRadius,
        leftSpacing: leftSpacing,
        disablePress: disablePress,
        totalHeight: totalHeight,
        height: height,
        setHeight: setHeight,
        getBarHeight: getBarHeight,
        getPosition: getPosition,
        getLowestPosition: getLowestPosition,
        lowestBarPosition: lowestBarPosition,
        getStackBorderRadii: getStackBorderRadii,
        tooltipProps: tooltipProps,
        renderTooltipConditions: (_j = props.renderTooltipConditions) !== null && _j !== void 0 ? _j : BarDefaults.renderTooltipConditions
    };
};
